app/scripts/components/common/map/style-generators/wms-timeseries.tsx

```diff
@@ -0,0 +1,172 @@
+import qs from 'qs';
+import { RasterSourceSpecification, RasterLayerSpecification } from 'mapbox-gl';
+
+import useFitBbox from '../hooks/use-fit-bbox';
+import useGeneratorParams from '../hooks/use-generator-params';
+import useMapStyle from '../hooks/use-map-style';
+import { BaseGeneratorParams } from '../types';
+
+import { useWMS } from '$components/common/map/style-generators/hooks';
+import { ActionStatus } from '$utils/status';
+
+import { userTzDate2utcString } from '$utils/date';
+
+// @NOTE: Some WMS Layers don't have a timestamp
+export interface MapLayerWMSProps extends BaseGeneratorParams {
+  id: string;
+  date?: Date;
+  stacCol: string;
+  sourceParams?: Record<string, any>;
+  stacApiEndpoint?: string;
+  zoomExtent?: number[];
+  onStatusChange?: (result: { status: ActionStatus; id: string }) => void;
+}
+
+interface ArcPaintLayerProps extends BaseGeneratorParams {
+  id: string;
+  sourceParams?: Record<string, any>;
+  date?: Date;
+  zoomExtent?: number[];
+  wmsUrl: string;
+  bounds: [number, number, number, number];
+}
+
+export function ArcPaintLayer(props: ArcPaintLayerProps) {
+  const {
+    id,
+    date,
+    sourceParams,
+    zoomExtent,
+    wmsUrl,
+    bounds,
+    generatorOrder,
+    hidden,
+    opacity
+  } = props;
+
+  const { updateStyle } = useMapStyle();
+
+  const [minZoom] = zoomExtent ?? [0, 20];
+
+  const generatorId = 'wms-' + id;
+
+  const generatorParams = useGeneratorParams({
+    generatorOrder,
+    hidden,
+    opacity
+  });
+
+  // Generate Mapbox GL layers and sources for raster layer
+  //
+  const haveSourceParamsChanged = useMemo(
+    () => JSON.stringify(sourceParams),
+    [sourceParams]
+  );
+
+  useEffect(
+    () => {
+      if (!wmsUrl) return;
+
+      const tileParams = qs.stringify({
+        // these are mostly gonna be same for wms layers, but users can override using sourceParams
+        format: 'image/png',
+        service: 'WMS',
+        request: 'GetMap',
+        transparent: 'true',
+        width: '256',
+        height: '256',
+        version: '1.3.0',
+        crs: 'EPSG:3857',
+        ...(date && { date: userTzDate2utcString(date) }),
+        ...sourceParams
+      });
+
+      const wmsSource: RasterSourceSpecification = {
+        type: 'raster',
+        tiles: [`${wmsUrl}?${tileParams}&bbox={bbox-epsg-3857}`],
+        tileSize: 256
+      };
+
+      const rasterOpacity = typeof opacity === 'number' ? opacity / 100 : 1;
+
+      const wmsLayer: RasterLayerSpecification = {
+        id: id,
+        type: 'raster',
+        source: id,
+        layout: {
+          visibility: hidden ? 'none' : 'visible'
+        },
+        paint: {
+          'raster-opacity': hidden ? 0 : rasterOpacity,
+          'raster-opacity-transition': {
+            duration: 320
+          }
+        },
+        minzoom: minZoom,
+        metadata: {
+          id,
+          layerOrderPosition: 'raster',
+          wmsTileUrl: `${wmsUrl}?${tileParams}`
```

Comment:
I am really hoping to consolidate raster-based paint layers (This work is in process: https://github.com/NASA-IMPACT/veda-ui/pull/1580) With that in mind, can you tell me why wms needed its own paint layer? I can see some metadata like this one, but is there anymore reason to have paint layer for wms layer?

Reply:
you're right, we can reuse `RasterPaintLayer`
updated [here](https://github.com/NASA-IMPACT/veda-ui/pull/1565/commits/2c95c4b8eb577192f765f2db85a246a774354984)